crates/re_viewer/build.rs

```diff
@@ -1,4 +1,212 @@
+                }
+            }
+        }
+
+        let thumbnail_dimensions = get!(example, thumbnail_dimensions);
+
+        Ok(Self {
+            title: get!(example, title),
+            description: get!(example, description),
+            tags: get!(example, tags),
+            demo_url: format!("{base_url}/examples/{}/", example.name),
+            rrd_url: format!("{base_url}/examples/{}/data.rrd", example.name),
+            thumbnail: Thumbnail {
+                url: get!(example, thumbnail),
+                width: thumbnail_dimensions[0],
+                height: thumbnail_dimensions[1],
+            },
+            name: example.name,
+        })
+    }
+}
+
+struct Example {
+    name: String,
+    readme: Frontmatter,
+}
+
+fn examples() -> Result<Vec<Example>> {
+    let mut examples = vec![];
+    for folder in std::fs::read_dir("../../examples/python")? {
+        let folder = folder?;
+        let metadata = folder.metadata()?;
+        let name = folder.file_name().to_string_lossy().to_string();
+        let readme = folder.path().join("README.md");
+        if metadata.is_dir() && readme.exists() {
+            let readme = parse_frontmatter(readme)?;
+            let Some(readme) = readme else { continue };
+            if readme.build_args.is_none() {
+                continue;
+            }
+            examples.push(Example { name, readme });
+        }
+    }
+    Ok(examples)
+}
+
+fn parse_frontmatter<P: AsRef<Path>>(path: P) -> Result<Option<Frontmatter>> {
+    let path = path.as_ref();
+    let content = std::fs::read_to_string(path)?;
+    let Some(content) = content.strip_prefix("---\n") else {
+        return Ok(None);
+    };
+    let Some(end) = content.find("---") else {
+        bail!("{:?} has invalid frontmatter", path);
+    };
+    Ok(Some(serde_yaml::from_str(&content[..end]).map_err(
+        |e| {
+            error!(
+                "failed to read {:?}: {e}",
+                path.parent().unwrap().file_name().unwrap()
+            )
+        },
+    )?))
+}
+
+const MANIFEST_PATH: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/data/examples_manifest.json");
+
+fn write_examples_manifest() -> Result<()> {
+    let mut manifest = vec![];
+    for example in examples()? {
+        manifest.push(ManifestEntry::try_from(example)?);
+    }
+    re_build_tools::write_file_if_necessary(
+        MANIFEST_PATH,
+        serde_json::to_string_pretty(&manifest)?.as_bytes(),
+    )?;
+    Ok(())
+}
+
+fn write_examples_manifest_if_necessary() {
+    if !re_build_tools::is_tracked_env_var_set("IS_IN_RERUN_WORKSPACE")
+        || re_build_tools::is_tracked_env_var_set("RERUN_IS_PUBLISHING")
+    {
+        return;
+    }
+    re_build_tools::rerun_if_changed_or_doesnt_exist(MANIFEST_PATH);
+
+    if let Err(err) = write_examples_manifest() {
+        panic!("{err}");
+    }
+}
+
 fn main() {
     re_build_tools::rebuild_if_crate_changed("re_viewer");
```

Comment:
This will re-run this build.rs script if anything in `re_viewer` or any of its dependancies changes. Isn't that a bit too often? I thought only the CI needed to run this at all? And it need to run it _always_,

Reply:
The build script has to run locally so that we can commit the `examples_manifest.json` file when it changes. That way users can build the `re_viewer` crate without being in the workspace. We need to trigger a run of the build script on any change to the `examples` directory, but I don't know what the best way to detect if examples were added/changed/removed is. Always running it solves that, I think?

I tried profiling the build script, and maybe it won't turn into an issue. Before this PR, the script takes ~350ms to run. Calling `write_examples_manifest` added ~500us for me. On CI it will add a bit more, because it also has to run `git` a few times.

Reply:
Oh, I missed where I was.

> We need to trigger a run of the build script on any change to the examples directory, but I don't know what the best way to detect if examples were added/changed/removed is

For changed/removed, use `re_build_tools::rerun_if_changed` for each file you read. It does `println!("cargo:rerun-if-changed=…` internally.

Detecting added files… I don't know how to do.

Reply:
> Always running it solves that, I think?

If you want to force the `build.rs` file to _always_ run (which you may want to on CI), there is an ugly trick: `println!("cargo:rerun-if-changed=path/to/file/that/doesnt/exist");`

Reply:
It should already always run on CI, because all the crates are built from scratch